include/vigra/threadpool.hxx

```diff
@@ -0,0 +1,328 @@
+/*                                                                      */
+/*    Permission is hereby granted, free of charge, to any person       */
+/*    obtaining a copy of this software and associated documentation    */
+/*    files (the "Software"), to deal in the Software without           */
+/*    restriction, including without limitation the rights to use,      */
+/*    copy, modify, merge, publish, distribute, sublicense, and/or      */
+/*    sell copies of the Software, and to permit persons to whom the    */
+/*    Software is furnished to do so, subject to the following          */
+/*    conditions:                                                       */
+/*                                                                      */
+/*    The above copyright notice and this permission notice shall be    */
+/*    included in all copies or substantial portions of the             */
+/*    Software.                                                         */
+/*                                                                      */
+/*    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND    */
+/*    EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES   */
+/*    OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND          */
+/*    NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT       */
+/*    HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,      */
+/*    WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING      */
+/*    FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR     */
+/*    OTHER DEALINGS IN THE SOFTWARE.                                   */
+/*                                                                      */
+/************************************************************************/
+#ifndef VIGRA_THREADPOOL_HXX
+#define VIGRA_THREADPOOL_HXX
+
+#include <functional>
+#include <thread>
+#include <atomic>
+#include <vector>
+#include <future>
+#include <mutex>
+#include <queue>
+#include <condition_variable>
+#include <stdexcept>
+#include <cmath>
+
+
+namespace vigra
+{
+
+
+
+class ThreadPool {
+public:
+
+    ThreadPool(size_t);
+
+    template<class F>
+    std::future<typename std::result_of<F(int)>::type>  enqueueReturning(F&& f) ;
+
+    template<class F>
+    void enqueue(F&& f) ;
+
+    ~ThreadPool();
+
+    void waitFinished()
+    {
+        std::unique_lock<std::mutex> lock(queue_mutex);
+        conditionF.wait(lock, [this](){ return tasks.empty() && (busy == 0); });
+    }
+
+    size_t nThreads()const{
+        return workers.size();
+    }
+private:
+    // need to keep track of threads so we can join them
+    std::vector< std::thread > workers;
+
+    // the task queue
+    std::queue< std::function<void(int)> > tasks;
+    
+    // synchronization
+    std::mutex queue_mutex;
+    std::condition_variable condition;
+    std::condition_variable conditionF;
+
+    bool stop;
```

Comment:
It does not matter whether `stop` is atomic or not. The lock is needed for the condition variable that wakes up waiting threads.